File: dbflow/src/main/java/com/raizlabs/android/dbflow/config/FlowManager.java
package com.raizlabs.android.dbflow.config;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.converter.TypeConverter;
import com.raizlabs.android.dbflow.runtime.ModelNotifier;
import com.raizlabs.android.dbflow.runtime.TableNotifierRegister;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.migration.Migration;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.raizlabs.android.dbflow.structure.BaseModelView;
import com.raizlabs.android.dbflow.structure.BaseQueryModel;
import com.raizlabs.android.dbflow.structure.InstanceAdapter;
import com.raizlabs.android.dbflow.structure.InvalidDBConfiguration;
import com.raizlabs.android.dbflow.structure.Model;
import com.raizlabs.android.dbflow.structure.ModelAdapter;
import com.raizlabs.android.dbflow.structure.ModelViewAdapter;
import com.raizlabs.android.dbflow.structure.QueryModelAdapter;
import com.raizlabs.android.dbflow.structure.RetrievalAdapter;
import com.raizlabs.android.dbflow.structure.database.DatabaseWrapper;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Description: The main entry point into the generated database code. It uses reflection to look up
 * and construct the generated database holder class used in defining the structure for all databases
 * used in this application.
 */
public class FlowManager {

    private static class GlobalDatabaseHolder extends DatabaseHolder {

        private boolean initialized = false;

        public void add(DatabaseHolder holder) {
            databaseDefinitionMap.putAll(holder.databaseDefinitionMap);
            databaseNameMap.putAll(holder.databaseNameMap);
            typeConverters.putAll(holder.typeConverters);
            databaseClassLookupMap.putAll(holder.databaseClassLookupMap);
            initialized = true;
        }

        public boolean isInitialized() {
            return initialized;
        }
    }

    static FlowConfig config;

    private static GlobalDatabaseHolder globalDatabaseHolder = new GlobalDatabaseHolder();

    private static HashSet<Class<? extends DatabaseHolder>> loadedModules = new HashSet<>();

    private static final String DEFAULT_DATABASE_HOLDER_NAME = "GeneratedDatabaseHolder";

    private static final String DEFAULT_DATABASE_HOLDER_PACKAGE_NAME =
        FlowManager.class.getPackage().getName();

    private static final String DEFAULT_DATABASE_HOLDER_CLASSNAME =
        DEFAULT_DATABASE_HOLDER_PACKAGE_NAME + "." + DEFAULT_DATABASE_HOLDER_NAME;

    /**
     * Returns the table name for the specific model class
     *
     * @param table The class that implements {@link Model}
     * @return The table name, which can be different than the {@link Model} class name
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static String getTableName(Class<?> table) {
        ModelAdapter modelAdapter = getModelAdapterOrNull(table);
        String tableName = null;
        if (modelAdapter == null) {
            ModelViewAdapter modelViewAdapter = getModelViewAdapterOrNull(table);
            if (modelViewAdapter != null) {
                tableName = modelViewAdapter.getViewName();
            } else {
                throwCannotFindAdapter("ModelAdapter/ModelViewAdapter", table);
            }
        } else {
            tableName = modelAdapter.getTableName();
        }
        return tableName;
    }

    /**
     * @param databaseName The name of the database. Will throw an exception if the database doesn't exist.
     * @param tableName    The name of the table in the DB.
     * @return The associated table class for the specified name.
     */
    public static Class<?> getTableClassForName(String databaseName, String tableName) {
        DatabaseDefinition databaseDefinition = getDatabase(databaseName);
        Class<?> modelClass = databaseDefinition.getModelClassForName(tableName);
        if (modelClass == null) {
            modelClass = databaseDefinition.getModelClassForName(QueryBuilder.quote(tableName));
            if (modelClass == null) {
                throw new IllegalArgumentException(String.format("The specified table %1s was not found. " +
                        "Did you forget to add the @Table annotation and point it to %1s?",
                    tableName, databaseName));
            }
        }
        return modelClass;
    }

    /**
     * @param table The table to lookup the database for.
     * @return the corresponding {@link DatabaseDefinition} for the specified model
     */
    @NonNull
    public static DatabaseDefinition getDatabaseForTable(Class<?> table) {
        checkDatabaseHolder();
        DatabaseDefinition databaseDefinition = globalDatabaseHolder.getDatabaseForTable(table);
        if (databaseDefinition == null) {
            throw new InvalidDBConfiguration("Model object: " + table.getName() +
                " is not registered with a Database. " + "Did you forget an annotation?");
        }
        return databaseDefinition;
    }

    @NonNull
    public static DatabaseDefinition getDatabase(Class<?> databaseClass) {
        checkDatabaseHolder();
        DatabaseDefinition databaseDefinition = globalDatabaseHolder.getDatabase(databaseClass);
        if (databaseDefinition == null) {
            throw new InvalidDBConfiguration("Database: " + databaseClass.getName() + " is not a registered Database. " +
                "Did you forget the @Database annotation?");
        }
        return databaseDefinition;
    }

    @NonNull
    public static DatabaseWrapper getWritableDatabaseForTable(Class<?> table) {
        return getDatabaseForTable(table).getWritableDatabase();
    }

    /**
     * @param databaseName The name of the database. Will throw an exception if the database doesn't exist.
     * @return the {@link DatabaseDefinition} for the specified database
     */
    @NonNull
    public static DatabaseDefinition getDatabase(String databaseName) {
        checkDatabaseHolder();
        DatabaseDefinition database = globalDatabaseHolder.getDatabase(databaseName);

        if (database != null) {
            return database;
        }

        throw new InvalidDBConfiguration("The specified database" + databaseName + " was not found. " +
            "Did you forget the @Database annotation?");
    }

    @NonNull
    public static DatabaseWrapper getWritableDatabase(String databaseName) {
        return getDatabase(databaseName).getWritableDatabase();
    }

    @NonNull
    public static DatabaseWrapper getWritableDatabase(Class<?> databaseClass) {
        return getDatabase(databaseClass).getWritableDatabase();
    }

    /**
     * Loading the module Database holder via reflection.
     * <p>
     * It is assumed FlowManager.init() is called by the application that uses the
     * module database. This method should only be called if you need to load databases
     * that are part of a module. Building once will give you the ability to add the class.
     */
    public static void initModule(Class<? extends DatabaseHolder> generatedClassName) {
        loadDatabaseHolder(generatedClassName);
    }

    public static FlowConfig getConfig() {
        if (config == null) {
            throw new IllegalStateException("Configuration is not initialized. " +
                "Please call init(FlowConfig) in your application class.");
        }
        return config;
    }

    /**
     * @return The database holder, creating if necessary using reflection.
     */
    protected static void loadDatabaseHolder(Class<? extends DatabaseHolder> holderClass) {
        if (loadedModules.contains(holderClass)) {
            return;
        }

        try {
            // Load the database holder, and add it to the global collection.
            DatabaseHolder dbHolder = holderClass.newInstance();

            if (dbHolder != null) {
                globalDatabaseHolder.add(dbHolder);

                // Cache the holder for future reference.
                loadedModules.add(holderClass);
            }
        } catch (Throwable e) {
            e.printStackTrace();
            throw new ModuleNotFoundException("Cannot load " + holderClass, e);
        }
    }

    /**
     * Resets all databases and associated files.
     */
    public static void reset() {
        Set<Map.Entry<Class<?>, DatabaseDefinition>> entrySet = globalDatabaseHolder.databaseClassLookupMap.entrySet();
        for (Map.Entry<Class<?>, DatabaseDefinition> value : entrySet) {
            value.getValue().reset(getContext());
        }
        globalDatabaseHolder.reset();
        loadedModules.clear();
    }

    /**
     * Will throw an exception if this class is not initialized yet in {@link #init(FlowConfig)}
     *
     * @return The shared context.
     */
    @NonNull
    public static Context getContext() {
        if (config == null) {
            throw new IllegalStateException("You must provide a valid FlowConfig instance. " +
                "We recommend calling init() in your application class.");
        }
        return config.getContext();
    }

    /**
     * Helper method to simplify the {@link #init(FlowConfig)}. Use {@link #init(FlowConfig)} to provide
     * more customization.
     *
     * @param context - should be application context, but not necessary as we retrieve it anyways.
     */
    public static void init(@NonNull Context context) {
        init(new FlowConfig.Builder(context).build());
    }

    /**
     * Initializes DBFlow, loading the main application Database holder via reflection one time only.
     * This will trigger all creations, updates, and instantiation for each database defined.
     *
     * @param flowConfig The configuration instance that will help shape how DBFlow gets constructed.
     */
    public static void init(@NonNull FlowConfig flowConfig) {
        FlowManager.config = flowConfig;

        try {
            //noinspection unchecked
            Class<? extends DatabaseHolder> defaultHolderClass = (Class<? extends DatabaseHolder>) Class.forName(DEFAULT_DATABASE_HOLDER_CLASSNAME);
            loadDatabaseHolder(defaultHolderClass);
        } catch (ModuleNotFoundException e) {
            // Ignore this exception since it means the application does not have its
            // own database. The initialization happens because the application is using
            // a module that has a database.
            FlowLog.log(FlowLog.Level.W, e.getMessage());
        } catch (ClassNotFoundException e) {
            // warning if a library uses DBFlow with module support but the app you're using doesn't support it.
            FlowLog.log(FlowLog.Level.W, "Could not find the default GeneratedDatabaseHolder");
        }

        if (flowConfig.databaseHolders() != null && !flowConfig.databaseHolders().isEmpty()) {
            for (Class<? extends DatabaseHolder> holder : flowConfig.databaseHolders()) {
                loadDatabaseHolder(holder);
            }
        }

        if (flowConfig.openDatabasesOnInit()) {
            List<DatabaseDefinition> databaseDefinitions = globalDatabaseHolder.getDatabaseDefinitions();
            for (DatabaseDefinition databaseDefinition : databaseDefinitions) {
                // triggers open, create, migrations.
                databaseDefinition.getWritableDatabase();
            }
        }
    }

    /**
     * @param objectClass A class with an associated type converter. May return null if not found.
     * @return The specific {@link TypeConverter} for the specified class. It defines
     * how the custom datatype is handled going into and out of the DB.
     */
    public static TypeConverter getTypeConverterForClass(Class<?> objectClass) {
        checkDatabaseHolder();
        return globalDatabaseHolder.getTypeConverterForClass(objectClass);
    }

    // region Getters

    /**
     * Release reference to context and {@link FlowConfig}
     */
    public static synchronized void destroy() {
        Set<Map.Entry<Class<?>, DatabaseDefinition>> entrySet =
            globalDatabaseHolder.databaseClassLookupMap.entrySet();
        for (Map.Entry<Class<?>, DatabaseDefinition> value : entrySet) {
            value.getValue().destroy(getContext());
        }

        config = null;

        // Reset the global database holder.
        globalDatabaseHolder = new GlobalDatabaseHolder();
        loadedModules.clear();
    }

    /**
     * @param modelClass The class that implements {@link Model} to find an adapter for.
     * @return The adapter associated with the class. If its not a {@link ModelAdapter},
     * it checks both the {@link ModelViewAdapter} and {@link QueryModelAdapter}.
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static <TModel> InstanceAdapter<TModel> getInstanceAdapter(Class<TModel> modelClass) {
        InstanceAdapter internalAdapter = getModelAdapterOrNull(modelClass);
        if (internalAdapter == null) {
            internalAdapter = getModelViewAdapterOrNull(modelClass);
            if (internalAdapter == null) {
                internalAdapter = getQueryModelAdapterOrNull(modelClass);
            }
        }

        if (internalAdapter == null) {
            throwCannotFindAdapter("InstanceAdapter", modelClass);
        }

        return internalAdapter;
    }

    /**
     * @param modelClass The class that implements {@link Model} to find an adapter for.
     * @return The adapter associated with the class. If its not a {@link ModelAdapter},
     * it checks both the {@link ModelViewAdapter} and {@link QueryModelAdapter}.
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static <TModel> RetrievalAdapter<TModel> getRetrievalAdapter(Class<TModel> modelClass) {
        RetrievalAdapter<TModel> retrievalAdapter = getModelAdapterOrNull(modelClass);
        if (retrievalAdapter == null) {
            retrievalAdapter = getModelViewAdapterOrNull(modelClass);
            if (retrievalAdapter == null) {
                retrievalAdapter = getQueryModelAdapterOrNull(modelClass);
            }
        }
        if (retrievalAdapter == null) {
            throwCannotFindAdapter("RetrievalAdapter", modelClass);
        }

        return retrievalAdapter;
    }


    /**
     * @param modelClass The class of the table
     * @param <TModel>   The class that implements {@link Model}
     * @return The associated model adapter (DAO) that is generated from a {@link Table} class. Handles
     * interactions with the database. This method is meant for internal usage only.
     * We strongly prefer you use the built-in methods associated with {@link Model} and {@link BaseModel}.
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static <TModel> ModelAdapter<TModel> getModelAdapter(Class<TModel> modelClass) {
        final ModelAdapter modelAdapter = getModelAdapterOrNull(modelClass);
        if (modelAdapter == null) {
            throwCannotFindAdapter("ModelAdapter", modelClass);
        }
        return modelAdapter;
    }

    /**
     * Returns the model view adapter for a SQLite VIEW. These are only created with the {@link com.raizlabs.android.dbflow.annotation.ModelView} annotation.
     *
     * @param modelViewClass The class of the VIEW
     * @param <TModelView>   The class that extends {@link BaseModelView}
     * @return The model view adapter for the specified model view.
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static <TModelView> ModelViewAdapter<TModelView> getModelViewAdapter(
        Class<TModelView> modelViewClass) {
        final ModelViewAdapter modelViewAdapter = getModelViewAdapterOrNull(modelViewClass);
        if (modelViewAdapter == null) {
            throwCannotFindAdapter("ModelViewAdapter", modelViewClass);
        }
        return modelViewAdapter;
    }

    /**
     * Returns the query model adapter for an undefined query. These are only created with the {@link TQueryModel} annotation.
     *
     * @param queryModelClass The class of the query
     * @param <TQueryModel>   The class that extends {@link BaseQueryModel}
     * @return The query model adapter for the specified model query.
     */
    @SuppressWarnings("unchecked")
    @NonNull
    public static <TQueryModel> QueryModelAdapter<TQueryModel> getQueryModelAdapter(
        Class<TQueryModel> queryModelClass) {
        final QueryModelAdapter queryModelAdapter = getQueryModelAdapterOrNull(queryModelClass);
        if (queryModelAdapter == null) {
            throwCannotFindAdapter("QueryModelAdapter", queryModelClass);
        }
        return queryModelAdapter;
    }

    @NonNull
    public static ModelNotifier getModelNotifierForTable(Class<?> table) {
        return getDatabaseForTable(table).getModelNotifier();
    }

    @NonNull
    public static TableNotifierRegister newRegisterForTable(Class<?> table) {
        return getModelNotifierForTable(table).newRegister();
    }

    @Nullable
    private static <T> ModelAdapter<T> getModelAdapterOrNull(Class<T> modelClass) {
        return FlowManager.getDatabaseForTable(modelClass).getModelAdapterForTable(modelClass);
    }

    @Nullable
    private static <T> ModelViewAdapter<T> getModelViewAdapterOrNull(Class<T> modelClass) {
        return FlowManager.getDatabaseForTable(modelClass).getModelViewAdapterForTable(modelClass);
    }

    @Nullable
    private static <T> QueryModelAdapter<T> getQueryModelAdapterOrNull(Class<T> modelClass) {
        return FlowManager.getDatabaseForTable(modelClass).getQueryModelAdapterForQueryClass(modelClass);
    }

    /**
     * @param databaseName The name of the database. Will throw an exception if the database doesn't exist.
     * @return The map of migrations for the specified database.
     */
    static Map<Integer, List<Migration>> getMigrations(String databaseName) {
        return getDatabase(databaseName).getMigrations();
    }

    /**
     * Checks a standard database helper for integrity using quick_check(1).
     *
     * @param databaseName The name of the database to check. Will thrown an exception if it does not exist.
     * @return true if it's integrity is OK.
     */
    public static boolean isDatabaseIntegrityOk(String databaseName) {
        return getDatabase(databaseName).getHelper().isDatabaseIntegrityOk();
    }

    private static void throwCannotFindAdapter(String type, Class<?> clazz) {
        throw new IllegalArgumentException("Cannot find " + type + " for " + clazz + ". Ensure " +
            "the class is annotated with proper annotation.");
    }

    private static void checkDatabaseHolder() {
        if (!globalDatabaseHolder.isInitialized()) {
            throw new IllegalStateException("The global database holder is not initialized. Ensure you call " +
                "FlowManager.init() before accessing the database.");
        }
    }

    // endregion

    /**
     * Exception thrown when a database holder cannot load the database holder
     * for a module.
     */
    public static class ModuleNotFoundException extends RuntimeException {
        public ModuleNotFoundException() {
        }

        public ModuleNotFoundException(String detailMessage) {
            super(detailMessage);
        }

        public ModuleNotFoundException(String detailMessage, Throwable throwable) {
            super(detailMessage, throwable);
        }

        public ModuleNotFoundException(Throwable throwable) {
            super(throwable);
        }
    }

}

File: dbflow-common/src/main/kotlin/com/raizlabs/dbflow5/config/FlowManager.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.raizlabs.dbflow5.config

import com.raizlabs.dbflow5.Synchronized
import com.raizlabs.dbflow5.adapter.ModelAdapter
import com.raizlabs.dbflow5.adapter.ModelViewAdapter
import com.raizlabs.dbflow5.adapter.QueryModelAdapter
import com.raizlabs.dbflow5.adapter.RetrievalAdapter
import com.raizlabs.dbflow5.converter.TypeConverter
import com.raizlabs.dbflow5.database.DBFlowDatabase
import com.raizlabs.dbflow5.migration.Migration
import com.raizlabs.dbflow5.quote
import com.raizlabs.dbflow5.runtime.ModelNotifier
import com.raizlabs.dbflow5.runtime.TableNotifierRegister
import com.raizlabs.dbflow5.structure.*
import kotlin.reflect.KClass

expect object FlowManager : FlowCommonManager

/**
 * Description: The main entry point into the generated database code. It uses reflection to look up
 * and construct the generated database holder class used in defining the structure for all databases
 * used in this application.
 */
abstract class FlowCommonManager {

    internal var config: FlowConfig? = null

    protected var globalDatabaseHolder = GlobalDatabaseHolder()

    protected val loadedModules = hashSetOf<KClass<out DatabaseHolder>>()

    private val DEFAULT_DATABASE_HOLDER_NAME = "GeneratedDatabaseHolder"

    private val DEFAULT_DATABASE_HOLDER_PACKAGE_NAME = "com.raizlabs.dbflow5.config"

    private val DEFAULT_DATABASE_HOLDER_CLASSNAME =
        "$DEFAULT_DATABASE_HOLDER_PACKAGE_NAME.$DEFAULT_DATABASE_HOLDER_NAME"

    protected class GlobalDatabaseHolder : DatabaseHolder() {

        var isInitialized = false
            private set

        fun add(holder: DatabaseHolder) {
            databaseDefinitionMap.putAll(holder.databaseDefinitionMap)
            databaseNameMap.putAll(holder.databaseNameMap)
            typeConverters.putAll(holder.typeConverters)
            databaseClassLookupMap.putAll(holder.databaseClassLookupMap)
            isInitialized = true
        }
    }

    /**
     * Returns the table name for the specific model class
     *
     * @param table The class that implements [Model]
     * @return The table name, which can be different than the [Model] class name
     */
    fun getTableName(table: KClass<*>): String {
        return getModelAdapterOrNull(table)?.tableName
            ?: getModelViewAdapterOrNull(table)?.viewName
            ?: throwCannotFindAdapter("ModelAdapter/ModelViewAdapter", table)
    }

    /**
     * @param databaseName The name of the database. Will throw an exception if the databaseForTable doesn't exist.
     * @param tableName    The name of the table in the DB.
     * @return The associated table class for the specified name.
     */
    fun getTableClassForName(databaseName: String, tableName: String): KClass<*> {
        val databaseDefinition = getDatabase(databaseName)
        return databaseDefinition.getModelClassForName(tableName)
            ?: databaseDefinition.getModelClassForName(tableName.quote())
            ?: throw IllegalArgumentException("The specified table $tableName was not found." +
                " Did you forget to add the @Table annotation and point it to $databaseName?")
    }

    /**
     * @param databaseClass The class of the database. Will throw an exception if the databaseForTable doesn't exist.
     * @param tableName     The name of the table in the DB.
     * @return The associated table class for the specified name.
     */

    fun getTableClassForName(databaseClass: KClass<out DBFlowDatabase>, tableName: String): KClass<*> {
        val databaseDefinition = getDatabase(databaseClass)
        return databaseDefinition.getModelClassForName(tableName)
            ?: databaseDefinition.getModelClassForName(tableName.quote())
            ?: throw IllegalArgumentException("The specified table $tableName was not found." +
                " Did you forget to add the @Table annotation and point it to $databaseClass?")
    }

    /**
     * @param table The table to lookup the database for.
     * @return the corresponding [DBFlowDatabase] for the specified model
     */

    fun getDatabaseForTable(table: KClass<*>): DBFlowDatabase {
        checkDatabaseHolder()
        return globalDatabaseHolder.getDatabaseForTable(table)
            ?: throw InvalidDBConfiguration("Model object: $table is not registered with a Database." +
                " Did you forget an annotation?")
    }

    @Suppress("UNCHECKED_CAST")

    fun <T : DBFlowDatabase> getDatabase(databaseClass: KClass<T>): T {
        checkDatabaseHolder()
        return globalDatabaseHolder.getDatabase(databaseClass) as? T
            ?: throw InvalidDBConfiguration("Database: $databaseClass is not a registered Database. " +
                "Did you forget the @Database annotation?")
    }


    fun getDatabaseName(database: KClass<out DBFlowDatabase>): String = getDatabase(database).databaseName

    /**
     * @param databaseName The name of the database. Will throw an exception if the databaseForTable doesn't exist.
     * @return the [DBFlowDatabase] for the specified database
     */

    fun getDatabase(databaseName: String): DBFlowDatabase {
        checkDatabaseHolder()
        return globalDatabaseHolder.getDatabase(databaseName)
            ?: throw InvalidDBConfiguration("The specified database $databaseName was not found. " +
                "Did you forget the @Database annotation?")
    }

    /**
     * Loading the module Database holder via reflection.
     *
     *
     * It is assumed FlowManager.init() is called by the application that uses the
     * module database. This method should only be called if you need to load databases
     * that are part of a module. Building once will give you the ability to add the class.
     */

    fun initModule(generatedClassName: KClass<out DatabaseHolder>) {
        loadDatabaseHolder(generatedClassName)
    }


    fun getConfig(): FlowConfig = config
        ?: throw IllegalStateException("Configuration is not initialized. " +
            "Please call init(FlowConfig) in your application class.")

    /**
     * @return The database holder, creating if necessary using reflection.
     */

    internal abstract fun loadDatabaseHolder(holderClass: KClass<out DatabaseHolder>)

    /**
     * Resets all databases and associated files.
     */
    @Synchronized

    fun reset() {
        globalDatabaseHolder.databaseClassLookupMap.values.forEach { it.reset() }
        globalDatabaseHolder.reset()
        loadedModules.clear()
    }

    /**
     * Close all DB files and resets [FlowConfig] and the [GlobalDatabaseHolder]. Brings
     * DBFlow back to initial application state.
     */
    @Synchronized

    fun close() {
        globalDatabaseHolder.databaseClassLookupMap.values.forEach { it.close() }
        config = null
        globalDatabaseHolder = GlobalDatabaseHolder()
        loadedModules.clear()
    }

    /**
     * Initializes DBFlow, loading the main application Database holder via reflection one time only.
     * This will trigger all creations, updates, and instantiation for each database defined.
     *
     * @param flowConfig The configuration instance that will help shape how DBFlow gets constructed.
     */

    protected fun initialize(flowConfig: FlowConfig) {
        config = config?.merge(flowConfig) ?: flowConfig

        loadDefaultHolderClass(DEFAULT_DATABASE_HOLDER_CLASSNAME)

        flowConfig.databaseHolders.forEach { loadDatabaseHolder(it) }

        if (flowConfig.openDatabasesOnInit) {
            globalDatabaseHolder.databaseDefinitions.forEach {
                // triggers open, create, migrations.
                it.writableDatabase
            }
        }
    }

    abstract fun loadDefaultHolderClass(className: String)

    /**
     * @param objectClass A class with an associated type converter. May return null if not found.
     * @return The specific [TypeConverter] for the specified class. It defines
     * how the custom datatype is handled going into and out of the DB.
     */

    fun getTypeConverterForClass(objectClass: KClass<*>): TypeConverter<*, *>? {
        checkDatabaseHolder()
        return globalDatabaseHolder.getTypeConverterForClass(objectClass)
    }

    // region Getters

    /**
     * Release reference to context and [FlowConfig]
     */

    @Synchronized
    fun destroy() {
        globalDatabaseHolder.databaseClassLookupMap.values.forEach { it.destroy() }
        config = null
        // Reset the global database holder.
        globalDatabaseHolder = GlobalDatabaseHolder()
        loadedModules.clear()
    }

    /**
     * @param modelClass The class that implements [Model] to find an adapter for.
     * @return The adapter associated with the class. If its not a [ModelAdapter],
     * it checks both the [ModelViewAdapter] and [QueryModelAdapter].
     */

    fun <T : Any> getRetrievalAdapter(modelClass: KClass<T>): RetrievalAdapter<T> {
        var retrievalAdapter: RetrievalAdapter<T>? = getModelAdapterOrNull(modelClass)
        if (retrievalAdapter == null) {
            retrievalAdapter = getModelViewAdapterOrNull(modelClass)
            if (retrievalAdapter == null) {
                retrievalAdapter = getQueryModelAdapterOrNull(modelClass)
            }
        }
        return retrievalAdapter ?: throwCannotFindAdapter("RetrievalAdapter", modelClass)
    }


    /**
     * @param modelClass The class of the table
     * @param [T]   The class that implements [Model]
     * @return The associated model adapter (DAO) that is generated from a [Table] class. Handles
     * interactions with the database. This method is meant for internal usage only.
     * We strongly prefer you use the built-in methods associated with [Model] and [BaseModel].
     */

    fun <T : Any> getModelAdapter(modelClass: KClass<T>): ModelAdapter<T> =
        getModelAdapterOrNull(modelClass) ?: throwCannotFindAdapter("ModelAdapter", modelClass)

    /**
     * Returns the model view adapter for a SQLite VIEW. These are only created with the [com.raizlabs.android.dbflow.annotation.ModelView] annotation.
     *
     * @param modelViewClass The class of the VIEW
     * @param [T]  The class that extends [BaseModelView]
     * @return The model view adapter for the specified model view.
     */
    fun <T : Any> getModelViewAdapter(modelViewClass: KClass<T>): ModelViewAdapter<T> =
        getModelViewAdapterOrNull(modelViewClass)
            ?: throwCannotFindAdapter("ModelViewAdapter", modelViewClass)

    /**
     * Returns the query model adapter for an undefined cursor. These are only created with the [T] annotation.
     *
     * @param queryModelClass The class of the query
     * @param [T]  The class that extends [BaseQueryModel]
     * @return The query model adapter for the specified model cursor.
     */

    fun <T : Any> getQueryModelAdapter(queryModelClass: KClass<T>): QueryModelAdapter<T> =
        getQueryModelAdapterOrNull(queryModelClass)
            ?: throwCannotFindAdapter("QueryModelAdapter", queryModelClass)


    fun getModelNotifierForTable(table: KClass<*>): ModelNotifier =
        getDatabaseForTable(table).getModelNotifier()

    fun newRegisterForTable(table: KClass<*>): TableNotifierRegister =
        getModelNotifierForTable(table).newRegister()

    private fun <T : Any> getModelAdapterOrNull(modelClass: KClass<T>): ModelAdapter<T>? =
        getDatabaseForTable(modelClass).getModelAdapterForTable(modelClass)

    private fun <T : Any> getModelViewAdapterOrNull(modelClass: KClass<T>): ModelViewAdapter<T>? =
        getDatabaseForTable(modelClass).getModelViewAdapterForTable(modelClass)

    private fun <T : Any> getQueryModelAdapterOrNull(modelClass: KClass<T>): QueryModelAdapter<T>? =
        getDatabaseForTable(modelClass).getQueryModelAdapterForQueryClass(modelClass)

    /**
     * @param databaseName The name of the database. Will throw an exception if the databaseForTable doesn't exist.
     * @return The map of migrations for the specified database.
     */

    internal fun getMigrations(databaseName: String): Map<Int, List<Migration>> =
        getDatabase(databaseName).migrations

    /**
     * Checks a standard database helper for integrity using quick_check(1).
     *
     * @param databaseName The name of the database to check. Will thrown an exception if it does not exist.
     * @return true if it's integrity is OK.
     */
    fun isDatabaseIntegrityOk(databaseName: String) = getDatabase(databaseName).openHelper.isDatabaseIntegrityOk

    private fun throwCannotFindAdapter(type: String, clazz: KClass<*>): Nothing =
        throw IllegalArgumentException("Cannot find $type for $clazz. Ensure the class is annotated with proper annotation.")

    private fun checkDatabaseHolder() {
        if (!globalDatabaseHolder.isInitialized) {
            throw IllegalStateException("The global databaseForTable holder is not initialized. " +
                "Ensure you call FlowManager.init() before accessing the databaseForTable.")
        }
    }

    // endregion

    /**
     * Exception thrown when a database holder cannot load the databaseForTable holder
     * for a module.
     */
    class ModuleNotFoundException(detailMessage: String, throwable: Throwable)
        : RuntimeException(detailMessage, throwable)

}

/**
 * Easily get access to its [DBFlowDatabase] directly.
 */
inline fun <reified T : DBFlowDatabase> database(): T = database(T::class)

inline fun <T : DBFlowDatabase> database(kClass: KClass<T>): T = FlowManager.getDatabase(kClass)

/**
 * Easily get access to its [DBFlowDatabase] directly.
 */
inline fun <T : DBFlowDatabase, R> database(kClass: KClass<T>, f: T.() -> R): R = database(kClass).f()

inline fun <reified T : DBFlowDatabase> database(f: T.() -> Unit): T = database(T::class).apply(f)

inline fun <T : Any, R> databaseForTable(kClass: KClass<T>, f: DBFlowDatabase.() -> R): R = databaseForTable(kClass).f()

inline fun <T : Any> databaseForTable(kClass: KClass<T>): DBFlowDatabase = FlowManager.getDatabaseForTable(kClass)

inline fun <reified T : Any> databaseForTable(f: DBFlowDatabase.() -> Unit): DBFlowDatabase = databaseForTable(T::class).apply(f)

/**
 * Easily get access to its [DBFlowDatabase] directly.
 */
inline fun <reified T : Any> databaseForTable(): DBFlowDatabase = databaseForTable(T::class)

/**
 * Easily get its table name.
 */
inline fun <reified T : Any> tableName(): String = FlowManager.getTableName(T::class)

/**
 * Easily get its [ModelAdapter].
 */
inline fun <reified T : Any> modelAdapter() = FlowManager.getModelAdapter(T::class)

inline val <T : Any> KClass<T>.modelAdapter
    get() = FlowManager.getModelAdapter(this)

inline val <T : Any> KClass<T>.retrievalAdapter
    get() = FlowManager.getRetrievalAdapter(this)

/**
 * Easily get its [QueryModelAdapter].
 */
inline fun <reified T : Any> queryModelAdapter() = FlowManager.getQueryModelAdapter(T::class)

inline val <T : Any> KClass<T>.queryModelAdapter
    get() = FlowManager.getQueryModelAdapter(this)

/**
 * Easily get its [ModelViewAdapter]
 */
inline fun <reified T : Any> modelViewAdapter() = FlowManager.getModelViewAdapter(T::class)

inline val <T : Any> KClass<T>.modelViewAdapter
    get() = FlowManager.getModelViewAdapter(this)
